server/src/main/java/org/elasticsearch/index/mapper/Mapping.java

```diff
@@ -63,13 +67,23 @@ public int compare(Mapper o1, Mapper o2) {
         this.metadataMappersMap = unmodifiableMap(metadataMappersMap);
         this.metadataMappersByName = unmodifiableMap(metadataMappersByName);
         this.meta = meta;
+        try {
+            this.mappingSource = new CompressedXContent(this, XContentType.JSON, ToXContent.EMPTY_PARAMS);
+        } catch (Exception e) {
+            throw new ElasticsearchGenerationException("failed to serialize source for type [" + root.name() + "]", e);
+        }
+
```

Comment:
Do we actually need to cache this? From looking at the various callers it seems that it's used in a) tests, b) assertions during updates and c) to merge mappings in the master - in other words, where it counts we only actually use it once.  Maybe the simplest thing to do is just return `new CompressedXContent(this.mapping(), XContentType.JSON, ToXContent.EMPTY_PARAMS)` from `DocumentMapper.mappingSource()`?

Reply:
you make a good point, I find this odd and I was also not too sure what the best way to proceed was. Let me double check